newIDE/app/src/Leaderboard/LeaderboardProvider.js

```diff
@@ -0,0 +1,406 @@
+      const fetchedEntries:
+        | LeaderboardEntry[]
+        | LeaderboardExtremePlayerScore[] = data.entries;
+
+      if (data.nextPageUri) {
+        dispatch({ type: 'SET_NEXT_PAGE_URI', payload: data.nextPageUri });
+      }
+
+      let entriesToDisplay: LeaderboardDisplayData[] = [];
+      if (displayOnlyBestEntry) {
+        entriesToDisplay = fetchedEntries.map(entry =>
+          // $FlowIgnore
+          extractExtremeScoreDisplayData(entry)
+        );
+      } else {
+        entriesToDisplay = fetchedEntries.map(entry =>
+          // $FlowIgnore
+          extractEntryDisplayData(entry)
+        );
+      }
+      dispatch({ type: 'SET_ENTRIES', payload: entriesToDisplay });
+    },
+    [currentLeaderboardId, displayOnlyBestEntry, gameId]
+  );
+
+  const selectLeaderboard = React.useCallback((leaderboardId: string) => {
+    dispatch({ type: 'SELECT_LEADERBOARD', payload: leaderboardId });
+  }, []);
+
+  const setDisplayOnlyBestEntry = React.useCallback((newValue: boolean) => {
+    dispatch({ type: 'CHANGE_DISPLAY_ONLY_BEST_ENTRY', payload: newValue });
+  }, []);
+
+  const updateLeaderboard = async (attributes: {|
+    name?: string,
+    sort?: LeaderboardSortOption,
+    playerUnicityDisplayChoice?: LeaderboardPlayerUnicityDisplayOption,
+  |}) => {
+    if (!currentLeaderboardId) return;
+    if (attributes.sort) dispatch({ type: 'PURGE_NAVIGATION' }); // When changing playerUnicityDisplayChoice, it will change the displayOnlyBestEntry state variable, which will purge navigation.
+    const updatedLeaderboard = await doUpdateLeaderboard(
+      authenticatedUser,
+      gameId,
+      currentLeaderboardId,
+      attributes
+    );
+    if (!updatedLeaderboard) return;
+
+    dispatch({
+      type: 'UPDATE_OR_CREATE_LEADERBOARD',
+      payload: updatedLeaderboard,
+    });
+
+    if (attributes.sort) await fetchEntries();
+  };
+
+  const resetLeaderboard = async () => {
+    if (!currentLeaderboardId) return;
+    dispatch({ type: 'PURGE_NAVIGATION' });
+    const updatedLeaderboard = await doResetLeaderboard(
+      authenticatedUser,
+      gameId,
+      currentLeaderboardId
+    );
+    if (!updatedLeaderboard) return;
+
+    dispatch({
+      type: 'UPDATE_OR_CREATE_LEADERBOARD',
+      payload: updatedLeaderboard,
+    });
+    fetchEntries();
+  };
+
+  const deleteLeaderboard = async () => {
+    if (!currentLeaderboardId || !leaderboardsByIds) return;
+    dispatch({ type: 'PURGE_NAVIGATION' });
+    await doDeleteLeaderboard(authenticatedUser, gameId, currentLeaderboardId);
+    dispatch({ type: 'REMOVE_LEADERBOARD', payload: currentLeaderboardId });
+  };
+
+  const deleteLeaderboardEntry = async (entryId: string) => {
+    if (!currentLeaderboardId) return;
+    await doDeleteLeaderboardEntry(
+      authenticatedUser,
+      gameId,
+      currentLeaderboardId,
+      entryId
+    );
+    fetchEntries({ uri: pageIndex > 0 ? mapPageIndexToUri[pageIndex] : null });
+  };
+
+  // --- Navigation ---
```

Comment:
Nitpicking: if you feel like you need to write a comment to separate things, it means that your code is becoming too long (though I agree it's hard to reduce).
What about making here a custom hook "useLeaderboardEntriesNavigation" that returns navigateToNextPage (and previous and first page)?